# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/RemoteTextureMap.h
# Commit: 21be69951c96
# Full Hash: 21be69951c962d3833d7117f56c2909b20b3d841
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2024-01-14 09:31:25
# Regressor Bug: 1874534
# File Overlap Count: 2
# Description:
#   Bug 1874534 - Fix DrawTargetWebgl and RemoteTextureMap context loss handling. r=aosmond
#   
#   When a context loss occurs on DrawTargetWebgl, this may result in a fallback TextureData
#   being created. Each of these are currently managed by two different RemoteTextureOwnerClients.
#   This is not really safe at all.
# ==============================================================================

diff -r 0930954b46bb -r 21be69951c96 gfx/layers/RemoteTextureMap.h
--- a/gfx/layers/RemoteTextureMap.h	Sat Jan 13 18:00:25 2024 +0000
+++ b/gfx/layers/RemoteTextureMap.h	Sat Jan 13 18:19:24 2024 +0000
@@ -187,6 +187,9 @@
   std::deque<Wait> mWaits;
 };
 
+typedef std::unordered_set<RemoteTextureOwnerId, RemoteTextureOwnerId::HashFn>
+    RemoteTextureOwnerIdSet;
+
 /**
  * A class provides API for remote texture owners.
  */
@@ -202,8 +205,9 @@
   void UnregisterTextureOwner(const RemoteTextureOwnerId aOwnerId);
   void UnregisterAllTextureOwners();
   void ClearRecycledTextures();
-  void NotifyContextLost();
-  void NotifyContextRestored();
+  void NotifyContextLost(const RemoteTextureOwnerIdSet* aOwnerIds = nullptr);
+  void NotifyContextRestored(
+      const RemoteTextureOwnerIdSet* aOwnerIds = nullptr);
   void PushTexture(const RemoteTextureId aTextureId,
                    const RemoteTextureOwnerId aOwnerId,
                    UniquePtr<TextureData>&& aTextureData);
@@ -243,8 +247,7 @@
  protected:
   ~RemoteTextureOwnerClient();
 
-  std::unordered_set<RemoteTextureOwnerId, RemoteTextureOwnerId::HashFn>
-      mOwnerIds;
+  RemoteTextureOwnerIdSet mOwnerIds;
   RefPtr<RemoteTextureRecycleBin> mSharedRecycleBin;
 };
 
@@ -288,24 +291,16 @@
       const RefPtr<RemoteTextureRecycleBin>& aRecycleBin = nullptr);
   void UnregisterTextureOwner(const RemoteTextureOwnerId aOwnerId,
                               const base::ProcessId aForPid);
-  void UnregisterTextureOwners(
-      const std::unordered_set<RemoteTextureOwnerId,
-                               RemoteTextureOwnerId::HashFn>& aOwnerIds,
-      const base::ProcessId aForPid);
+  void UnregisterTextureOwners(const RemoteTextureOwnerIdSet& aOwnerIds,
+                               const base::ProcessId aForPid);
 
   void ClearRecycledTextures(
-      const std::unordered_set<RemoteTextureOwnerId,
-                               RemoteTextureOwnerId::HashFn>& aOwnerIds,
-      const base::ProcessId aForPid,
+      const RemoteTextureOwnerIdSet& aOwnerIds, const base::ProcessId aForPid,
       const RefPtr<RemoteTextureRecycleBin>& aRecycleBin = nullptr);
-  void NotifyContextLost(
-      const std::unordered_set<RemoteTextureOwnerId,
-                               RemoteTextureOwnerId::HashFn>& aOwnerIds,
-      const base::ProcessId aForPid);
-  void NotifyContextRestored(
-      const std::unordered_set<RemoteTextureOwnerId,
-                               RemoteTextureOwnerId::HashFn>& aOwnerIds,
-      const base::ProcessId aForPid);
+  void NotifyContextLost(const RemoteTextureOwnerIdSet& aOwnerIds,
+                         const base::ProcessId aForPid);
+  void NotifyContextRestored(const RemoteTextureOwnerIdSet& aOwnerIds,
+                             const base::ProcessId aForPid);
 
   // Get remote texture's TextureHost for RemoteTextureHostWrapper.
   //